### ISSUE TITLE
Add a logger sink for training progress, as `compresso` 0.1.7 has

### ISSUE THREAD

**Issue by zombak79** [zombak79/compresso-recsys#23]
## Problem

Training progress is reported only through `tqdm`, gated on `show_progress`. That works in a notebook and nowhere
else. A service embedding the library has to turn it off, and then gets nothing.

Concretely, `python-models` (PYT-405) wraps the whole zoo behind one family member and forces `show_progress=False`
for two reasons, both of which hold independently:

- `tqdm` writes carriage returns to stdout. The service's logging goes through its own `RemoteLogger` to a collector;
  a progress bar there is noise at best, and at worst it is the only thing in the log.
- `show_progress` is read on the **prediction** path too (`ELSATrainer.predict`, `evaluate_recommender`), and
  `recommend()`/`predict()` route through it. A served model with `show_progress=True` draws a bar *per request*.

So the flag cannot be left on, and once it is off a fit is completely silent. The first production target is ELSA on
19,338,738 users x 132,860 items (811,056,679 nnz), **CPU-only**, where a fit is estimated at 0.8-4 h depending on
`latent_dim`/`batch_size`/`max_output`. Between `Fitting ELSATrainer on ...` and `Fit complete in Xs` there is no
output at all. The incumbent it has to beat (SANSA) logs loss and max-residual throughout its 1.02-1.08 h fit, so the
new model is strictly worse to operate.

## Why a consumer-side workaround is not enough

Every trainer already accumulates per-epoch records in `self.history` and persists them in the checkpoint, so the
consumer can log those after `fit()` returns — and `python-models` now does.

That is not sufficient, because the records are **per epoch**. At 19.3M users the right setting is `epochs=1` (one
epoch there is 19.3M gradient updates; the documented `epochs=10` recipe is calibrated on datasets ~140x smaller in
users). With `epochs=1` the history has exactly one record, written at the end — 18,885 batches of silence at
`batch_size=1024`, then a single line.

Intra-epoch progress cannot be reached from outside the library: `_run_epoch` drives the bar per batch and there is no
other hook. Only a library-side sink fixes it.

## Reference implementation

`compresso` 0.1.7 solved this for `TopKSAETrainer` — see `_Reporter` in `src/compresso/trainers/saetrainer.py`. Its
design decisions are the ones worth porting, not just the parameter:

- **Duck-typed sink.** `logger: Any | None`, requiring only `.info(str)`. Works with `logging.Logger`, a service's own
  logger, or a shim — no dependency on `logging`, which `compresso-recsys` currently does not import anywhere in `src/`.
- **A logger wins over the bar unconditionally**: `show_progress = bool(show_progress) and logger is None`. Both carry
  the same numbers and tqdm has nowhere good to draw them without a tty, so passing a logger should not require also
  remembering to disable progress.
- **Resolved per call, not held on the trainer.** A logger describes the job that is running, not the model: a trainer
  fitted under a job logger may later be transformed under another, or none. It also scopes the failure latch to one
  call and keeps two threads sharing a trainer out of each other's reporting.
- **`log_every_n_steps`** for intra-epoch granularity, meaningful only when a logger is present. This is the part the
  consumer-side workaround cannot replicate.
- **A raising logger latches off rather than ending the call**, surfacing the first failure as a warning. Hours of
  training must not be lost to a broken log handler.
- **A prefix**, so lines are attributable: `f"[{prefix}] {message}"`.
- **An `_INHERIT` sentinel** for per-call arguments, so "not passed" is distinguishable from an explicit `None`.

## Scope

Six trainers default `show_progress: bool = True` and keep `self.history`:

| class | config default |
|---|---|
| `ELSATrainer` | `elsa.py:202` |
| `MultDAETrainer` | `mult_dae.py:59` |
| `MultVAETrainer` | `mult_vae.py:62` |
| `SimpleGPTTrainer` | `simple_gpt.py:161` |
| `SimpleRNNTrainer` | `simple_rnn.py:101` |
| `TEASERGDTrainer` | `teaser_gd.py:270` |

Also worth routing through the same sink:

- `evaluation.py:306` — `evaluate_recommender`'s tqdm, which is on the prediction path.
- `elsa.py:887` — a bare `print()` of forced-rewind stats during compressed training. In a service this goes to stdout
  and bypasses the log collector entirely.
- `builder.py:97,109` — `print(f"[compresso-recsys] {message}")`. Already prefixed, so it is the same idea done
  informally; folding it into the reporter would make it configurable.
- `datasets/amazon2023.py:284,346` — lower priority, dataset download is not a service path.

## Proposed API

```python
# constructor: the job-level default
ELSATrainer(config, logger=None)

# per call: overrides, distinguishing "not passed" from an explicit None
model.fit(interactions, *, logger=_INHERIT, show_progress=_INHERIT)
model.predict(source, *, logger=_INHERIT, show_progress=_INHERIT, ...)
```

with `log_every_n_steps` on the config (or the reporter), defaulting to something that yields a handful of lines per
epoch rather than per batch.

## Acceptance criteria

- Passing a `logger` produces per-epoch lines and, with `log_every_n_steps > 0`, intra-epoch lines, with **no tqdm
  output**, without also having to set `show_progress=False`.
- Nothing is written to stdout/stderr by a trainer when a logger is supplied.
- A logger whose `.info` raises does not abort training; the first failure is warned about once.
- The existing notebook behaviour is unchanged when no logger is passed.
- An object exposing only `.info(str)` is accepted (no `logging.Logger` isinstance check).

## Notes

- This is the same fix `compresso` took in 0.1.7 (`2c283a3c`), which let PYT-397 delete a polling thread over
  `trainer.history`. The equivalent workaround is now in `python-models` for `compresso-recsys`, and it can be deleted
  the same way once this lands.
- Version coordination: `python-models` pins `compresso-recsys==0.3.0`, so this needs a release before the consumer
  side can be simplified. No rush — the workaround is adequate for per-epoch reporting; only intra-epoch progress is
  blocked on it.

**Comment by zombak79** [comment]
implemented in 0.3.1
